### PR TITLE
Add api section about interoperabilty

### PR DESCRIPTION
**Pull request by tingerrr**
This PR adds a section to the API part which explains different kinds of considerations to take to make a package API more interoperable.

Right now, this mainly contains a section about state, counter and label keys.

I think there's more to be researched and said, I want to look at polylux, touying and tablex for their cetz integration, as well as fletcher for its extension of cetz and escape hatch mechanism. All of these are good examples of interoperability that is not yet written down here.

### CODE REVIEW

**Comment by Jollywatt** [comment]
This is a good idea. I would suggest adding a concrete point about exports: try to always ensure non-trivial functions defined in you package are exported somehow, ideally via submodules.

For example, I'm currently reworking fletcher to fully support cetz coordinates, which would be much, much harder if cetz didn't export all its utility functions in `util` and `coordinate` submodules.

**Comment by tingerrr** [comment]
I'll take a look at this as well, it's definitely something I do too.

File: src/chapters/api/interoperability.typ [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+- should this item be public or private?
+- should this behavior be configurable?
+
+The choice to make an item private or part of a function's behavior hardcoded can negatively impact it's interoperability, but the opposite can lead to bloated and intimidating APIs.
+Depending on the target audience of your package you may chose one over the other.
+Developers may require APIs with more configurabe parameters while end users may just want a simple to use API that doesn't get in their way.
+
+== Visibility
+In order to provide other package authors with the ability to write extensions to your package they must usually be able to access cerain internals of your package.
```
Comment (jamesrswift) [inline src/chapters/api/interoperability.typ]: I assume you mean "certain"

File: src/chapters/api/interoperability.typ [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+The internals of your package can be exposed in different ways.
+They can be exposed as public items which are undocumented or marked as unstable.
+Or they can be exposed fully documented and stable but marked as internal or advanced.
+How and if they are exposed depends on your package's use cases and target audience.
+
+#wbox[
+  At the time of writing this Typst 0.11.1 doesn't support controlling the visiblity of items directly.
+]
+
+To mark an item private, consider simply prefixing it with an underscore like `_internal-func`.
```
Comment (jamesrswift) [inline src/chapters/api/interoperability.typ:27]: Might be worth mentioning how the `tidy` package can omit documentation for internals prefixed with an underscore, so they can still be documented in the source file without showing up in the manual. That might be too much though

File: src/chapters/api/interoperability.typ
Comment (jamesrswift) [inline src/chapters/api/interoperability.typ:1]: Is it worth issuing guidance on how "examples" are shown in readme files (i.e., when showing an example for a package, don't star-import it so as not to unnecessarily encourage namespace pollution)

**Review by jamesrswift — APPROVED** [review]
LGTM. Some small nitpicks but I don't feel too strongly about them so do feel free to ignore

File: src/chapters/api/interoperability.typ (hunk 2 above)
Comment (tingerrr) [reply src/chapters/api/interoperability.typ:27]: I would consider this an implementation detail of tidy, so I'll keep this out of the guidelines.

It's up to the project to decide whether such an item is documented or not.

File: src/chapters/api/interoperability.typ
Comment (tingerrr) [reply src/chapters/api/interoperability.typ:1]: That's a good idea, I'll add an example of that.

**Pull request merged**